File: ang-app/src/app/management/mod_cpfl/client/client.service.ts
import { Client } from './client';
import { BaseCrudService } from '../../../../app_base/basecrud.service';

export class ClientService extends BaseCrudService {

	getUrl(){
		return this.getBase() + '/clients';
	}

	getUrlForCache(){
		return this.getUrl();
	}

	getEmptyObject(conditions) :Object{
		var ownerId: number = 
			parseInt('0' + this.storageService.localStorageGetItem('_ownerId_' + this.getAppId()));
		return new Client(0,conditions,
		                  null,null,null,null,null,null,null,null,null,null,null,null,
		                  null,null,null,null,null,null,null,'null',
		                  ownerId,
		                  this.storageService.localStorageGetItem('_token_' + this.getAppId()),null);
	}

	getAll(page,rows,conditions): Promise<Client[]> {
		return new Promise<Client[]>(r => r(super.getAll(page,rows,this.getEmptyObject(conditions))));
	}

	load(id): Promise<Client> {
		return new Promise<Client>(r => r(super.load(id)));
	} 

	mergeAnotherInObject(object,anotherObject){
		object.a1_name = super.getChangedValue(anotherObject.a1_name,object.a1_name);
		object.a2_type = super.getChangedValue(anotherObject.a2_type,object.a2_type);
		object.a3_cpf = super.getChangedValue(anotherObject.a3_cpf,object.a3_cpf);
		object.a4_cnpj = super.getChangedValue(anotherObject.a4_cnpj,object.a4_cnpj);
		object.a5_birthdate = super.getChangedValue(anotherObject.a5_birthdate,object.a5_birthdate);
		object.a6_doctype = super.getChangedValue(anotherObject.a6_doctype,object.a6_doctype);
		object.a7_document = super.getChangedValue(anotherObject.a7_document,object.a7_document);
		object.a8_gender = super.getChangedValue(anotherObject.a8_gender,object.a8_gender);
		object.a9_email = super.getChangedValue(anotherObject.a9_email,object.a9_email);
		object.a10_phone = super.getChangedValue(anotherObject.a10_phone,object.a10_phone);
		object.a11_cep = super.getChangedValue(anotherObject.a11_cep,object.a11_cep);
		object.a12_uf = super.getChangedValue(anotherObject.a12_uf,object.a12_uf);
		object.a13_city = super.getChangedValue(anotherObject.a13_city,object.a13_city);
		object.a14_street = super.getChangedValue(anotherObject.a14_street,object.a14_street);
		object.a15_number = super.getChangedValue(anotherObject.a15_number,object.a15_number);
		object.a16_compl1type = super.getChangedValue(anotherObject.a16_compl1type,object.a16_compl1type);
		object.a17_compl1desc = super.getChangedValue(anotherObject.a17_compl1desc,object.a17_compl1desc);
		object.a18_compl2type = super.getChangedValue(anotherObject.a18_compl2type,object.a18_compl2type);
		object.a19_compl2desc = super.getChangedValue(anotherObject.a19_compl2desc,object.a19_compl2desc); 
		object.a5_birthdateLabel = 
			super.getChangedValue(anotherObject.a5_birthdateLabel,object.a5_birthdateLabel);
		object.ownerId = super.getChangedValue(anotherObject.ownerId,object.ownerId);
		return object;
	}

}

File: ex_app/lib/mod_cpfl/client/client.service.ex
defmodule ExApp.ClientService do

  use ExApp.BaseServiceSecure
  alias ExApp.App.DAOService
  alias ExApp.DateUtil
  alias ExApp.MapUtil
  alias ExApp.NumberUtil
  alias ExApp.ResultSetHandler
  alias ExApp.Client 
  alias ExApp.ConsumerunitService
  alias ExApp.SolicitationService
  
  
  def loadById(id) do
    sql = """
	      select id, a1_name,
          a2_type,
          a3_cpf,
          a4_cnpj,
          a5_birthdate,
          a6_doctype,
          a7_document,
          a8_gender,
          a9_email,
          a10_phone,
          a11_cep,
          a12_uf,
          a13_city,
          a14_street,
          a15_number,
          a16_compl1type,
          a17_compl1desc,
          a18_compl2type,
          a19_compl2desc, ownerId from client where id = ? limit 1
	      """
	resultset = DAOService.load(sql,[id])
	cond do
      (nil == resultset or resultset.num_rows == 0) -> nil
      true -> getClient(resultset)
    end
  end
  
  def create(paramValues) do
    sql = """
          insert into client(a1_name,
          a2_type,
          a3_cpf,
          a4_cnpj,
          a5_birthdate,
          a6_doctype,
          a7_document,
          a8_gender,
          a9_email,
          a10_phone,
          a11_cep,
          a12_uf,
          a13_city,
          a14_street,
          a15_number,
          a16_compl1type,
          a17_compl1desc,
          a18_compl2type,
          a19_compl2desc,ownerId,created_at) values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)
	      """
    DAOService.insert(sql,paramValues ++ [DateUtil.getNowToSql(0,false,false)])
  end
  
  def update(id,paramValues) do
    sql = """
          update client set a1_name = ?,
          a2_type = ?,
          a3_cpf = ?,
          a4_cnpj = ?,
          a5_birthdate = ?,
          a6_doctype = ?,
          a7_document = ?,
          a8_gender = ?,
          a9_email = ?,
          a10_phone = ?,
          a11_cep = ?,
          a12_uf = ?,
          a13_city = ?,
          a14_street = ?,
          a15_number = ?,
          a16_compl1type = ?,
          a17_compl1desc = ?,
          a18_compl2type = ?,
          a19_compl2desc = ?, updated_at = ? where id = ?
	      """
    DAOService.update(sql,paramValues ++ [DateUtil.getNowToSql(0,false,false),id])
  end
  
  def updateDependencies(id,client) do
    clientName = MapUtil.get(client,:a1_name)
    clientCpf = MapUtil.get(client,:a3_cpf)
    clientCnpj = MapUtil.get(client,:a4_cnpj)
    ConsumerunitService.updateByClient(id,clientName,clientCpf,clientCnpj)
    SolicitationService.updateByClient(id,clientName,clientCpf,clientCnpj)
  end
  
  def loadAll(page,rows,conditions,deletedAt,_mapParams) do
    limit = cond do
      (page > 0 and rows > 0) -> " limit #{((page - 1) * rows)},#{rows}"
      true -> ""
    end
    sql = "select count(id) from client where #{deletedAt} #{conditions}"
    resultset = DAOService.load(sql,[])
    total = NumberUtil.toInteger(ResultSetHandler.getColumnValue(resultset,0,0))
    sql = """
          select id, a1_name,
          a2_type,
          a3_cpf,
          a4_cnpj,
          a5_birthdate,
          a6_doctype,
          a7_document,
          a8_gender,
          a9_email,
          a10_phone,
          a11_cep,
          a12_uf,
          a13_city,
          a14_street,
          a15_number,
          a16_compl1type,
          a17_compl1desc,
          a18_compl2type,
          a19_compl2desc, 
          ownerId from client 
          where #{deletedAt} #{conditions} 
          order by id asc
          #{limit}
          """
    resultset = DAOService.load(sql,[])
    cond do
      (nil == resultset or resultset.num_rows == 0) 
        -> [Client.new(0,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,0,0)]
      true -> parseResults(resultset,total,[],0) 
    end
  end
  
  def loadAllForPublic(_page,_rows,_conditions,_deletedAt,_mapParams) do
    [Client.new(0,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,0,0)]
  end 
  
  def delete(id) do
    DAOService.update("update client set deleted_at = ? where id = ?",
                      [DateUtil.getNowToSql(0,false,false),id])
  end
  
  def unDrop(id) do
    DAOService.update("update client set deleted_at = null where id = ?",[id])
  end
  
  def trullyDrop(id) do
    DAOService.delete("delete from client where id = ?",[id])
  end
  
  defp parseResults(resultset,totalRows,arrayMap,row) do
    cond do
      (nil == resultset or resultset.num_rows == 0 or row >= resultset.num_rows) -> arrayMap
      true -> parseResults(resultset,totalRows,
                           arrayMap ++ [getClient(resultset,row,totalRows)],row + 1)
    end
  end
  
  defp getClient(resultset,row \\ 0,totalRows \\ nil) do
    total = cond do
      (row == 0) -> totalRows
      true -> nil
    end
    Client.new(NumberUtil.toInteger(ResultSetHandler.getColumnValue(resultset,row,0)),
               ResultSetHandler.getColumnValue(resultset,row,1),
               ResultSetHandler.getColumnValue(resultset,row,2),
               ResultSetHandler.getColumnValue(resultset,row,3),
               ResultSetHandler.getColumnValue(resultset,row,4),
               ResultSetHandler.getColumnValue(resultset,row,5),
               ResultSetHandler.getColumnValue(resultset,row,6),
               ResultSetHandler.getColumnValue(resultset,row,7),
               ResultSetHandler.getColumnValue(resultset,row,8),
               ResultSetHandler.getColumnValue(resultset,row,9),
               ResultSetHandler.getColumnValue(resultset,row,10),
               ResultSetHandler.getColumnValue(resultset,row,11),
               ResultSetHandler.getColumnValue(resultset,row,12),
               ResultSetHandler.getColumnValue(resultset,row,13),
               ResultSetHandler.getColumnValue(resultset,row,14),
               ResultSetHandler.getColumnValue(resultset,row,15),
               ResultSetHandler.getColumnValue(resultset,row,16),
               ResultSetHandler.getColumnValue(resultset,row,17),
               ResultSetHandler.getColumnValue(resultset,row,18),
               ResultSetHandler.getColumnValue(resultset,row,19),
               NumberUtil.toInteger(ResultSetHandler.getColumnValue(resultset,row,20)),
               total)
  end
  
end